Java 19 compatibility

In Java 19 the signature of the Java_sun_nio_ch_FileChannelImpl_map0() function has changed with [JDK-8286671](https://bugs.openjdk.org/browse/JDK-8286671). This leads to the following exception when running on Java 19:
```
 Caused by: java.lang.AssertionError: java.lang.NoSuchMethodException: sun.nio.ch.FileChannelImpl.map0(int,long,long,boolean)
 	at net.openhft.chronicle.core.Jvm.getMethod0(Jvm.java:636)
 	at net.openhft.chronicle.core.Jvm.getMethod(Jvm.java:611)
 	at net.openhft.chronicle.core.OS.lambda$static$0(OS.java:55)
 	at net.openhft.chronicle.core.ClassLocal.computeValue(ClassLocal.java:54)
 	at java.base/java.lang.ClassValue.getFromHashMap(ClassValue.java:229)
 	at java.base/java.lang.ClassValue.getFromBackup(ClassValue.java:211)
 	at java.base/java.lang.ClassValue.get(ClassValue.java:117)
 	at net.openhft.chronicle.core.OS.map0(OS.java:420)
 	at net.openhft.chronicle.core.OS.map(OS.java:388)
```

--- a/src/main/java/net/openhft/chronicle/core/Bootstrap.java
+++ b/src/main/java/net/openhft/chronicle/core/Bootstrap.java
@@ -49,6 +49,7 @@ final class Bootstrap {
     static final boolean IS_JAVA_12_PLUS;
     static final boolean IS_JAVA_14_PLUS;
     static final boolean IS_JAVA_15_PLUS;
+    static final boolean IS_JAVA_19_PLUS;
 
     static {
         JVM_JAVA_MAJOR_VERSION = Bootstrap.getMajorVersion0();
@@ -56,6 +57,7 @@ final class Bootstrap {
         IS_JAVA_12_PLUS = JVM_JAVA_MAJOR_VERSION > 11;
         IS_JAVA_14_PLUS = JVM_JAVA_MAJOR_VERSION > 13;
         IS_JAVA_15_PLUS = JVM_JAVA_MAJOR_VERSION > 14;
+        IS_JAVA_19_PLUS = JVM_JAVA_MAJOR_VERSION > 18;
     }
 
     private Bootstrap() {

--- a/src/main/java/net/openhft/chronicle/core/Jvm.java
+++ b/src/main/java/net/openhft/chronicle/core/Jvm.java
@@ -325,12 +325,21 @@ public final class Jvm {
     }
 
     /**
-     * Returns if the major Java version is 14 or higher.
+     * Returns if the major Java version is 15 or higher.
      *
-     * @return if the major Java version is 14 or higher
+     * @return if the major Java version is 15 or higher
      */
     public static boolean isJava15Plus() {
         return IS_JAVA_15_PLUS;
+    }
+
+    /**
+     * Returns if the major Java version is 19 or higher.
+     *
+     * @return if the major Java version is 19 or higher
+     */
+    public static boolean isJava19Plus() {
+        return IS_JAVA_19_PLUS;
     }
 
     private static boolean is64bit0() {

--- a/src/main/java/net/openhft/chronicle/core/OS.java
+++ b/src/main/java/net/openhft/chronicle/core/OS.java
@@ -27,6 +27,7 @@ import javax.naming.TimeLimitExceededException;
 import java.io.*;
 import java.lang.invoke.MethodHandle;
 import java.lang.invoke.MethodHandles;
+import java.lang.reflect.Field;
 import java.lang.reflect.InvocationTargetException;
 import java.lang.reflect.Method;
 import java.net.DatagramSocket;
@@ -55,13 +56,15 @@ public final class OS {
     static final ClassLocal<MethodHandle> MAP0_MH = ClassLocal.withInitial(c -> {
         try {
             Method map0;
-            if (Jvm.isJava14Plus()) map0 = Jvm.getMethod(c, "map0", int.class, long.class, long.class, boolean.class);
+            if (Jvm.isJava19Plus()) map0 = Jvm.getMethod(c, "map0", FileDescriptor.class, int.class, long.class, long.class, boolean.class);
+            else if (Jvm.isJava14Plus()) map0 = Jvm.getMethod(c, "map0", int.class, long.class, long.class, boolean.class);
             else map0 = Jvm.getMethod(c, "map0", int.class, long.class, long.class);
             return MethodHandles.lookup().unreflect(map0);
         } catch (IllegalAccessException e) {
             throw new AssertionError(e);
         }
     });
+    private static final Field FD_FIELD = Jvm.getField(FileChannelImpl.class, "fd");
     private static final String TARGET = findTarget();
     private static final String USER_NAME = Jvm.getProperty("user.name");
     private static final int MAP_RO = 0;
@@ -418,7 +421,10 @@ public final class OS {
         try {
             // For now, access is assumed to be non-synchronous
             // TODO - Support passing/deducing synchronous flag externally
-            if (Jvm.isJava14Plus())
+            if (Jvm.isJava19Plus()) {
+                final FileDescriptor fd = (FileDescriptor) FD_FIELD.get(fileChannel);
+                return (long) map0.invokeExact((FileChannelImpl) fileChannel, fd, imode, start, size, false);
+            } else if (Jvm.isJava14Plus())
                 return (long) map0.invokeExact((FileChannelImpl) fileChannel, imode, start, size, false);
             else
                 return (long) map0.invokeExact((FileChannelImpl) fileChannel, imode, start, size);

